include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,378 @@
+            rw.VdotH = VdotH;
+            L = V.template reflect<reflect_wrapper>(rw);
+
+            cache = anisocache_type::createForReflection(localV, localH);
+        }
+        else    // fail if samples have invalid paths
+            L.makeInvalid();    // should check if sample direction is invalid
+
+        const vector3_type T = interaction.getT();
+        const vector3_type B = interaction.getB();
+        const vector3_type _N = interaction.getN();
+
+        return sample_type::create(L, T, B, _N);
+    }
+    template<typename C=bool_constant<IsBSDF> >
+    enable_if_t<C::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const scalar_type NdotV = localV.z;
+
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, NdotV);
+        fresnel::OrientedEtaRcps<monochrome_type> rcpEta = _f.getOrientedEtaRcps();
+
+        const vector3_type upperHemisphereV = ieee754::flipSignIfRHSNegative<vector3_type>(localV, hlsl::promote<vector3_type>(NdotV));
+        const vector3_type localH = ndf.generateH(upperHemisphereV, u.xy);
+        const scalar_type VdotH = hlsl::dot(localV, localH);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        assert(NdotV * VdotH > scalar_type(0.0));
+        const scalar_type reflectance = _f(hlsl::abs(VdotH))[0];
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
+
+        ray_dir_info_type V = interaction.getV();
+        const vector3_type _N = interaction.getN();
+        Refract<scalar_type> r = Refract<scalar_type>::create(V.getDirection(), H);
+        const scalar_type LdotH = hlsl::mix(VdotH, r.getNdotT(rcpEta.value2[0]), transmitted);
+        cache = anisocache_type::createPartial(VdotH, LdotH, hlsl::dot(_N, H), transmitted, rcpEta);
+
+        struct reflect_refract_wrapper  // so we don't recalculate LdotH
+        {
+            vector3_type operator()(const bool doRefract, const scalar_type rcpOrientedEta) NBL_CONST_MEMBER_FUNC
+            {
+                return rr(NdotTorR, rcpOrientedEta);
+            }
+            bxdf::ReflectRefract<scalar_type> rr;
+            scalar_type NdotTorR;
+        };
+        bxdf::ReflectRefract<scalar_type> rr;
+        rr.refract = r;
+        reflect_refract_wrapper rrw;
+        rrw.rr = rr;
+        rrw.NdotTorR = LdotH;
+        ray_dir_info_type L = V.template reflectRefract<reflect_refract_wrapper>(rrw, transmitted, rcpEta.value[0]);
+
+        const vector3_type T = interaction.getT();
+        const vector3_type B = interaction.getB();
+
+        // fail if samples have invalid paths
+        const scalar_type NdotL = scalar_type(2.0) * VdotH * localH.z - localV.z;
```

Comment:
this is not a valid computation of NdotL for the BSDF, this is the reflection equation